torch/distributed/_shard/checkpoint/filesystem.py

```diff
@@ -0,0 +1,142 @@
+import os
+import operator
+import pickle
+from typing import List, Optional, cast
+
+import torch
+from torch import Tensor
+import torch.distributed as dist
+from torch.futures import Future
+from pathlib import Path
+
+from .metadata import (
+    BytesReadRequest,
+    BytesWriteRequest,
+    Metadata,
+    TensorReadRequest,
+    TensorWriteRequest,
+)
+from .storage import StorageReader, StorageWriter
+
+def _get_file_storage_path(
+    storage_path: str,
+    storage_key: str
+) -> Path:
+    return (Path(storage_path) / storage_key)
+
+class FileSystemWriter(StorageWriter):
+    def __init__(self, path: str) -> None:
+        super().__init__()
+        self.path = path
+        os.makedirs(self.path, exist_ok=True)
+
+    def _check_for_bad_keys(self, requests):
+        if any([r for r in requests if r.storage_key == ".metadata"]):
+            raise ValueError("Cannot use key named '.metadata'")
```

Comment:
It feels wrong to allow some StorageWriter implementations to only implement a subset of the API.  It seems like it will be confusing if checkpoint saving can fail with some StorageWriter implementations but succeed with others.

In general I think it would be better to just consistently ensure that we can support any arbitrary keys.  If we really want to reject some keys then it seems like we should do that consistently at a higher layer in the code, so that we don't have different behavior across different StorageWriters.

Note that with the current FileSystemWriter implementation we should also be rejecting any keys containing `/`  (and depending on the platform, possibly `\`).  We also probably can't handle the empty string, `.`, or `..` correctly.

Reply:
Good point. I implemented a simple FQN -> storage key mapping that should be enough to enable downstream users to start experimenting with this API.

This mapping schema only your concerns around FileSystemWriter and is not generally useful beyond it. I'll work on a design to address it after this PR.